Prepare RemoteScrollingCoordinatorTransaction pieces for generated serialization

achristensen07 opened the pull request:
#### 08f8141737e23917c19660c0beaa9ea871e752e0
<pre>
Prepare RemoteScrollingCoordinatorTransaction pieces for generated serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=261622">https://bugs.webkit.org/show_bug.cgi?id=261622</a>
rdar://115573578

Reviewed by Simon Fraser.

This moves the manual C++ serialization code around so that each piece has a symmetric encode
and decode function that only encodes and decodes itself with no additional context.  To do this,
I needed to have the scrolling state nodes deserialize only the values they contain and then
attach to the scrolling state tree after all the nodes are deserialized.

* Source/WebCore/page/scrolling/ScrollingStateFixedNode.cpp:
(WebCore::ScrollingStateFixedNode::create):
(WebCore::ScrollingStateFixedNode::ScrollingStateFixedNode):
* Source/WebCore/page/scrolling/ScrollingStateFixedNode.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameHostingNode.cpp:
(WebCore::ScrollingStateFrameHostingNode::create):
(WebCore::ScrollingStateFrameHostingNode::ScrollingStateFrameHostingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameHostingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::create):
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
(WebCore::ScrollingStateFrameScrollingNode::isMainFrame const):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::ScrollingStateNode):
(WebCore::ScrollingStateNode::attachAfterDeserialization):
(WebCore::ScrollingStateNode::setChildren):
(WebCore::ScrollingStateNode::recurse):
(WebCore::ScrollingStateNode::setLayer):
(WebCore::ScrollingStateNode::parentPointersAreCorrect const):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollProxyNode.cpp:
(WebCore::ScrollingStateOverflowScrollProxyNode::create):
(WebCore::ScrollingStateOverflowScrollProxyNode::ScrollingStateOverflowScrollProxyNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollProxyNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::create):
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePositionedNode.cpp:
(WebCore::ScrollingStatePositionedNode::create):
(WebCore::ScrollingStatePositionedNode::ScrollingStatePositionedNode):
* Source/WebCore/page/scrolling/ScrollingStatePositionedNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.cpp:
(WebCore::ScrollingStateStickyNode::create):
(WebCore::ScrollingStateStickyNode::ScrollingStateStickyNode):
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::attachDeserializedNodes):
(WebCore::ScrollingStateTree::setDeserializedRootStateNode):
(WebCore::ScrollingStateTree::rootStateNode const):
(WebCore::ScrollingStateTree::insertNode):
(WebCore::ScrollingStateTree::clear):
(WebCore::ScrollingStateTree::addNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
(WebCore::ScrollingStateTree::rootStateNode const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;ScrollingStateNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateNode&gt;::decode):
(encodeNodeShared):
(decodeNodeShared):
(encodeScrollingStateScrollingNodeShared):
(decodeScrollingStateScrollingNodeShared):
(ArgumentCoder&lt;ScrollingStateFrameScrollingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateFrameScrollingNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStateFrameHostingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateFrameHostingNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStateOverflowScrollingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateOverflowScrollingNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStateOverflowScrollProxyNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateOverflowScrollProxyNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStateFixedNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateFixedNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStateStickyNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateStickyNode&gt;::decode):
(ArgumentCoder&lt;ScrollingStatePositionedNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStatePositionedNode&gt;::decode):
(ArgumentCoder&lt;WebCore::ScrollingStateTree&gt;::encode):
(ArgumentCoder&lt;WebCore::ScrollingStateTree&gt;::decode):
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::encode): Deleted.
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::decode): Deleted.
(encodeNodeAndDescendants): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):

Canonical link: <a href="https://commits.webkit.org/269648@main">https://commits.webkit.org/269648@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06a7f26866af9b1a143b2774e4e0b9823e9b328d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23179 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1266 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24300 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/25100 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21426 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23446 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2769 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23718 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22276 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23420 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/20097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25942 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/693 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/20981 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27142 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/21141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/21245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25013 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/678 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/18445 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/612 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1089 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2942 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/887 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/scrolling/ScrollingStateNode.h:
```diff
@@ -315,6 +315,8 @@ class ScrollingStateNode : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
     const Vector<Ref<ScrollingStateNode>>& children() const { return m_children; }
     Vector<Ref<ScrollingStateNode>> takeChildren() { return std::exchange(m_children, { }); }
+    WEBCORE_EXPORT void setChildren(Vector<Ref<ScrollingStateNode>>&&);
+    void recurse(const Function<void(ScrollingStateNode&)>&);
```
There exists `ScrollingTree::traverseScrollingTree()`. Maybe we should add `ScrollingStateTree:: traverse()` instead of this?

smfr commented on Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
```diff
@@ -67,8 +67,37 @@ ScrollingStateTree::~ScrollingStateTree() = default;
+    }
+}
+
+bool ScrollingStateTree::setDeserializedRootStateNode(RefPtr<ScrollingStateFrameScrollingNode>&& node)
```
The "Deserialized" bit is really a WK2 implementation detail which is a bit weird to see here. Can we pick a more generic name? Maybe `setRootStateNodeAfterReconstruction` or something.

achristensen07 committed:
Prepare RemoteScrollingCoordinatorTransaction pieces for generated serialization
https://bugs.webkit.org/show_bug.cgi?id=261622
rdar://115573578

Reviewed by Simon Fraser.

This moves the manual C++ serialization code around so that each piece has a symmetric encode
and decode function that only encodes and decodes itself with no additional context.  To do this,
I needed to have the scrolling state nodes deserialize only the values they contain and then
attach to the scrolling state tree after all the nodes are deserialized.

* Source/WebCore/page/scrolling/ScrollingStateFixedNode.cpp:
(WebCore::ScrollingStateFixedNode::create):
(WebCore::ScrollingStateFixedNode::ScrollingStateFixedNode):
* Source/WebCore/page/scrolling/ScrollingStateFixedNode.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameHostingNode.cpp:
(WebCore::ScrollingStateFrameHostingNode::create):
(WebCore::ScrollingStateFrameHostingNode::ScrollingStateFrameHostingNode):
* Source/WebCore/page/scrolling/ScrollingStateFrameHostingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.cpp:
(WebCore::ScrollingStateFrameScrollingNode::create):
(WebCore::ScrollingStateFrameScrollingNode::ScrollingStateFrameScrollingNode):
(WebCore::ScrollingStateFrameScrollingNode::isMainFrame const):
* Source/WebCore/page/scrolling/ScrollingStateFrameScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateNode.cpp:
(WebCore::ScrollingStateNode::ScrollingStateNode):
(WebCore::ScrollingStateNode::attachAfterDeserialization):
(WebCore::ScrollingStateNode::setChildren):
(WebCore::ScrollingStateNode::recurse):
(WebCore::ScrollingStateNode::setLayer):
(WebCore::ScrollingStateNode::parentPointersAreCorrect const):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollProxyNode.cpp:
(WebCore::ScrollingStateOverflowScrollProxyNode::create):
(WebCore::ScrollingStateOverflowScrollProxyNode::ScrollingStateOverflowScrollProxyNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollProxyNode.h:
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.cpp:
(WebCore::ScrollingStateOverflowScrollingNode::create):
(WebCore::ScrollingStateOverflowScrollingNode::ScrollingStateOverflowScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateOverflowScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStatePositionedNode.cpp:
(WebCore::ScrollingStatePositionedNode::create):
(WebCore::ScrollingStatePositionedNode::ScrollingStatePositionedNode):
* Source/WebCore/page/scrolling/ScrollingStatePositionedNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.cpp:
(WebCore::ScrollingStateStickyNode::create):
(WebCore::ScrollingStateStickyNode::ScrollingStateStickyNode):
* Source/WebCore/page/scrolling/ScrollingStateStickyNode.h:
* Source/WebCore/page/scrolling/ScrollingStateTree.cpp:
(WebCore::ScrollingStateTree::attachDeserializedNodes):
(WebCore::ScrollingStateTree::setDeserializedRootStateNode):
(WebCore::ScrollingStateTree::rootStateNode const):
(WebCore::ScrollingStateTree::insertNode):
(WebCore::ScrollingStateTree::clear):
(WebCore::ScrollingStateTree::addNode):
* Source/WebCore/page/scrolling/ScrollingStateTree.h:
(WebCore::ScrollingStateTree::rootStateNode const): Deleted.
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::commitTreeState):
(WebCore::ScrollingTree::updateTreeFromStateNodeRecursive):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder<ScrollingStateNode>::encode):
(ArgumentCoder<ScrollingStateNode>::decode):
(encodeNodeShared):
(decodeNodeShared):
(encodeScrollingStateScrollingNodeShared):
(decodeScrollingStateScrollingNodeShared):
(ArgumentCoder<ScrollingStateFrameScrollingNode>::encode):
(ArgumentCoder<ScrollingStateFrameScrollingNode>::decode):
(ArgumentCoder<ScrollingStateFrameHostingNode>::encode):
(ArgumentCoder<ScrollingStateFrameHostingNode>::decode):
(ArgumentCoder<ScrollingStateOverflowScrollingNode>::encode):
(ArgumentCoder<ScrollingStateOverflowScrollingNode>::decode):
(ArgumentCoder<ScrollingStateOverflowScrollProxyNode>::encode):
(ArgumentCoder<ScrollingStateOverflowScrollProxyNode>::decode):
(ArgumentCoder<ScrollingStateFixedNode>::encode):
(ArgumentCoder<ScrollingStateFixedNode>::decode):
(ArgumentCoder<ScrollingStateStickyNode>::encode):
(ArgumentCoder<ScrollingStateStickyNode>::decode):
(ArgumentCoder<ScrollingStatePositionedNode>::encode):
(ArgumentCoder<ScrollingStatePositionedNode>::decode):
(ArgumentCoder<WebCore::ScrollingStateTree>::encode):
(ArgumentCoder<WebCore::ScrollingStateTree>::decode):
(ArgumentCoder<ScrollingStateScrollingNode>::encode): Deleted.
(ArgumentCoder<ScrollingStateScrollingNode>::decode): Deleted.
(encodeNodeAndDescendants): Deleted.
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteScrollingCoordinatorProxyIOS.mm:
(WebKit::RemoteScrollingCoordinatorProxyIOS::connectStateNodeLayers):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::connectStateNodeLayers):

Canonical link: https://commits.webkit.org/269648@main
